### PR TITLE
Category feature

### PR DESCRIPTION
**Pull request by Vadym-Ozi**

### CODE REVIEW

File: src/main/java/example/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    @PostMapping("/new")
+    @Operation(summary = "Create a new category")
+    public CategoryDto createCategory(@RequestBody @Valid CategoryDto categoryDto){
+        return categoryService.save(categoryDto);
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping
+    @Operation(summary = "Get list of all categories")
+    public List<CategoryDto> getAll() {
```
Comment (Elena-Bruyako) [inline src/main/java/example/controller/CategoryController.java:41]: Add Pageable

File: src/main/java/example/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+@RequestMapping("/categories")
+@Tag(name = "Category Management", description = "Endpoints for managing categories")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @PostMapping("/new")
+    @Operation(summary = "Create a new category")
+    public CategoryDto createCategory(@RequestBody @Valid CategoryDto categoryDto){
```
Comment (Elena-Bruyako) [inline src/main/java/example/controller/CategoryController.java:34]: CategoryDto should include id in response
create CategoryRequestDto

File: src/main/java/example/dto/category/CategoryDto.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package example.dto.category;
+
+import jakarta.validation.constraints.*;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class CategoryDto {
+    @Pattern(regexp = "^[A-Z][a-z]*$")
```
Comment (Elena-Bruyako) [inline src/main/java/example/dto/category/CategoryDto.java:10]: NotBlank?

File: src/main/java/example/mapper/BookMapper.java [hunk 4]
```diff
@@ -3,12 +3,29 @@
...
 public interface BookMapper {
     BookDto toDto(Book book);
 
-    Book toModel(CreateBookRequestDto requestDto);
+    Book toEntity(CreateBookRequestDto requestDto);
+
+    BookDtoWithoutCategoryIds toDtoWithoutCategories(Book book);
+
+    @AfterMapping
+    @EntityGraph(attributePaths = "categories")
```
Comment (Elena-Bruyako) [inline src/main/java/example/mapper/BookMapper.java:24]: Does this annotation work here?

File: src/main/java/example/service/impl/CategoryServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    public CategoryDto update(Long id, CategoryDto categoryDto) {
+        Category category = categoryMapper.toEntity(getById(id));
+        category.setId(id);
+        if (!(categoryDto.getName() == null|| categoryDto.getName().isBlank())) {
+            category.setName(categoryDto.getName());
+        }
+        if (!(categoryDto.getDescription() == null || categoryDto.getDescription().isBlank())) {
+            category.setDescription(categoryDto.getDescription());
+        }
+        return categoryMapper.toDto(categoryRepository.save(category));
```
Comment (Elena-Bruyako) [inline src/main/java/example/service/impl/CategoryServiceImpl.java:52]:
```suggestion
        Category category = repository.findById(id)
                .orElseThrow(() .....
                ));
        mapper.updateCategoryFromDto(categoryDto, category);
        return mapper.toDto(repository.save(category));
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/example/mapper/BookMapper.java (hunk 4 above)
Comment (Vadym-Ozi) [reply src/main/java/example/mapper/BookMapper.java:24]: nope, i was searching for a solution and forget to delete this.
fixed

File: src/main/java/example/controller/CategoryController.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+@RequestMapping("/categories")
+@Tag(name = "Category Management", description = "Endpoints for managing categories")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PreAuthorize("hasRole('USER')")
+    @PostMapping("/new")
+    @Operation(summary = "Create a new category")
+    public CategoryDto createCategory(@RequestBody @Valid CategoryDto categoryDto){
```
Comment (Elena-Bruyako) [inline src/main/java/example/controller/CategoryController.java:38]:
```suggestion
    public CategoryDto createCategory(@RequestBody @Valid CategoryRequestDto categoryDto){
```

File: src/main/java/example/controller/CategoryController.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import java.util.List;
+
+@RestController
```
Comment (Elena-Bruyako) [inline src/main/java/example/controller/CategoryController.java:27]:
```suggestion
@RestController
@Validated
```

if you use `@Posititve` for id

File: src/main/java/example/mapper/BookMapper.java (hunk 4 above)
Comment (Elena-Bruyako) [inline src/main/java/example/mapper/BookMapper.java:24]: The annotation is still here

File: src/main/java/example/mapper/CategoryMapper.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+public interface CategoryMapper {
+    CategoryDto toDto(Category category);
+
+    Category toEntity(CategoryDto categoryDTO);
+
+    @AfterMapping
+    default void updateCategoryFromDto(@MappingTarget CategoryRequestDto categoryRequestDto, Category category) {
+        category.setName(categoryRequestDto.getName());
+        category.setDescription(categoryRequestDto.getDescription());
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/example/mapper/CategoryMapper.java:21]:
```suggestion
    void updateCategoryFromDto(CategoryRequestDto categoryDto, @MappingTarget Category category);
```

File: src/main/java/example/model/Category.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Getter
+@Setter
+@Table(name = "categories")
+@SQLDelete(sql = "UPDATE categories SET is_deleted = true WHERE id = ?")
+@SQLRestriction("is_Deleted=false")
```
Comment (Elena-Bruyako) [inline src/main/java/example/model/Category.java:14]:
```suggestion
@SQLRestriction("is_deleted=false")
```

File: src/main/java/example/repository/book/BookRepository.java [hunk 10]
```diff
@@ -3,6 +3,10 @@
 import example.model.Book;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import java.util.List;
+
```
Comment (Elena-Bruyako) [inline src/main/java/example/repository/book/BookRepository.java:7]:
```suggestion
```

File: src/main/java/example/security/JwtAuthenticationFilter.java [hunk 11]
```diff
@@ -45,7 +44,7 @@ protected void doFilterInternal(
     private String getToken(HttpServletRequest request) {
         String bearerToken = request.getHeader(HEADER);
```
Comment (Elena-Bruyako) [inline src/main/java/example/security/JwtAuthenticationFilter.java:45]:
```suggestion
        String bearerToken = request.getHeader(HttpHeaders.AUTHORIZATION);
```

File: src/main/java/example/security/JwtAuthenticationFilter.java [hunk 12]
```diff
@@ -22,7 +22,6 @@ public class JwtAuthenticationFilter extends OncePerRequestFilter {
     private final UserDetailsService userDetailsService;
     private static final String HEADER = "Authorization";
```
Comment (Elena-Bruyako) [inline src/main/java/example/security/JwtAuthenticationFilter.java:23]:
```suggestion
```

File: src/main/java/example/service/impl/CategoryServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @Override
+    public CategoryDto getById(Long id) {
+        Category category = categoryRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can`t find category by id: " + id));
+        return categoryMapper.toDto(category);
+    }
+
+    @Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/example/service/impl/CategoryServiceImpl.java:43]:
```suggestion
```

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml [hunk 14]
```diff
@@ -1,6 +1,6 @@
 databaseChangeLog:
   - changeSet:
-      id: create-users-roles-table
+      id: create-users-roles-table.yaml
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml:3]: Why do you need yaml here?

File: src/main/resources/db/changelog/changes/06-insert-books.yaml [hunk 15]
```diff
@@ -1,6 +1,6 @@
 databaseChangeLog:
   - changeSet:
-      id: 02-insert-books
+      id: insert-books.yaml
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/06-insert-books.yaml:3]: same

File: src/main/resources/db/changelog/changes/05-insert-data-to-db.yaml [hunk 16]
```diff
@@ -1,16 +1,16 @@
 databaseChangeLog:
   - changeSet:
-      id: 01-create-book-table
+      id: insert-data-to-db.yaml
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/05-insert-data-to-db.yaml:3]: same

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]
good job

**Pull request closed without merging**